In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv(r'C:/Users/user/Downloads/Weather.csv')

In [3]:
df

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


In [5]:
df.describe().T

count        mean        std    min     25%     50%  \
Temp_C            8784.0    8.798144  11.687883 -23.30    0.10    9.30   
Dew Point Temp_C  8784.0    2.555294  10.883072 -28.50   -5.90    3.30   
Rel Hum_%         8784.0   67.431694  16.918881  18.00   56.00   68.00   
Wind Speed_km/h   8784.0   14.945469   8.688696   0.00    9.00   13.00   
Visibility_km     8784.0   27.664447  12.622688   0.20   24.10   25.00   
Press_kPa         8784.0  101.051623   0.844005  97.52  100.56  101.07   

                     75%     max  
Temp_C             18.80   33.00  
Dew Point Temp_C   11.80   24.40  
Rel Hum_%          81.00  100.00  
Wind Speed_km/h    20.00   83.00  
Visibility_km      25.00   48.30  
Press_kPa         101.59  103.65

In [6]:
df.shape #number of total number of rows and columns

(8784, 8)

In [7]:
df.head(10) #Top 10 rows

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   
5  1/1/2012 5:00    -1.4              -3.3         87                9   
6  1/1/2012 6:00    -1.5              -3.1         89                7   
7  1/1/2012 7:00    -1.4              -3.6         85                7   
8  1/1/2012 8:00    -1.4              -3.6         85                9   
9  1/1/2012 9:00    -1.3              -3.1         88               15   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog  
5            6.4     101.27                   Fog  
6            6.4     101.29                   Fog  
7            8.0     101.26                   Fog  
8            8.0     101.23                   Fog  
9            4.0     101.20                   Fog

In [8]:
df.tail(10) #Bottom 10 rows

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8774  12/31/2012 14:00    -3.4              -5.7         84                6   
8775  12/31/2012 15:00    -2.3              -4.6         84                9   
8776  12/31/2012 16:00    -1.4              -4.0         82               13   
8777  12/31/2012 17:00    -1.1              -3.3         85               19   
8778  12/31/2012 18:00    -1.3              -3.1         88               17   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather  
8774           11.3     100.57    Snow  
8775            9.7     100.47    Snow  
8776           12.9     100.40    Snow  
8777            9.7     100.30    Snow  
8778            9.7     100.19    Snow  
8779            9.7     100.13    Snow  
8780            9.7     100.03    Snow  
8781            4.8      99.95    Snow  
8782            9.7      99.91    Snow  
8783           11.3      99.89    Snow

In [9]:
df.isnull().sum() #Checking for null values

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

<h3> 1. Find all records where the weather was exactly clear </h3>

In [12]:
df['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [14]:
clear=df[df['Weather']=='clear']
print(clear)

Empty DataFrame
Columns: [Date/Time, Temp_C, Dew Point Temp_C, Rel Hum_%, Wind Speed_km/h, Visibility_km, Press_kPa, Weather]
Index: []


 This shows that 1326 records have clear weather. 

<h2> 2. Find the number of times the wind speed was exactly 4 km/hr  </h2>

In [14]:
df['Wind Speed_km/h'].value_counts()

Wind Speed_km/h
9     830
11    791
13    735
15    719
7     677
17    666
19    616
6     609
20    496
4     474
22    439
24    374
0     309
26    242
28    205
30    161
32    139
33     85
35     53
37     45
39     24
41     22
44     14
43     13
48     13
46     11
52      7
57      5
50      4
2       2
83      1
70      1
63      1
54      1
Name: count, dtype: int64

In [16]:
wind_speed = df[df['Wind Speed_km/h']== 4].shape[0]

In [18]:
print(f'Number of times the wind speed was exactly 4km/h4: {wind_speed}')

Number of times the wind speed was exactly 4km/h4: 474


<h2> 3. Check if there are any NULL values present in the dataset </h2>

In [15]:
df.isna().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

<h2> 4 Rename the column" weather" to "weather_condition." </h2>

In [23]:
df['Weather_condition']= df["Weather"]

In [24]:
df.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather',
       'Weather_condition'],
      dtype='object')

<h2>5 What is the mean visibility of the dataset</h2>

In [19]:
df['Visibility_km'].mean()

np.float64(27.664446721311478)

<h2>6. Find the number of records where the wind speed is greater than 24 km/hr and visibility is equal to 25km.</h2/

In [27]:
df['greater_than_24km'] = df['Wind Speed_km/h']>24 
df['visibility_equal25']=df['Visibility_km']=25 

In [25]:
filtered_df=df[(df['Wind Speed_km/h']> 24) & (df['Visibility_km'] == 25)]

In [26]:
filtered_df.count()

Date/Time            308
Temp_C               308
Dew Point Temp_C     308
Rel Hum_%            308
Wind Speed_km/h      308
Visibility_km        308
Press_kPa            308
Weather              308
Weather_condition    308
dtype: int64

<h2>7.What is the mean value of each column for each weather condition?<h2/>

In [30]:
df['Weather_condition'].describe()

count             8784
unique              50
top       Mainly Clear
freq              2106
Name: Weather_condition, dtype: object

<h2>8. Find all instances where the weather is clear and the relative humidity is greater than 50, or visibility is above 40.<h2/> 

In [31]:
clear = df[(df['Weather_condition'] == 'clear') & ((df['Rel Hum_%'] > 50) | 
                                                  (df['Visibility_km'] > 40 ))]

In [32]:
print("Instances where the weather is clearand the relative humidity is greater than 50 orr visibilityis above 40:")
print(clear)

Instances where the weather is clearand the relative humidity is greater than 50 orr visibilityis above 40:
Empty DataFrame
Columns: [Date/Time, Temp_C, Dew Point Temp_C, Rel Hum_%, Wind Speed_km/h, Visibility_km, Press_kPa, Weather, Weather_condition]
Index: []


<h2>9. Find the number of weather conditions that include snow<h2/>

In [29]:
df['Weather_condition'].value_counts()

Weather_condition
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle      

In [35]:
snow_conditions=df['Weather_condition'].str.contains('Snow', case=False).sum()

In [36]:
snow_conditions

np.int64(583)